sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -586,9 +584,139 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
 }
 
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def inputTypes: Seq[DataType] = {
+    // TODO: properly handle String Types with differing collations
+    val leftIsString = left.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    val rightIsString = right.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    if (leftIsString && rightIsString) {
+      Seq(left.dataType, right.dataType)
+    } else {
+      Seq(StringType, StringType)
+    }
+  }
+
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val lCollationID = left.dataType.asInstanceOf[StringType].collationId
+    val rCollationID = right.dataType.asInstanceOf[StringType].collationId
+    // TODO: ensure that lCollationID and rCollationID are equal (compatible)
+    val collationID = if (lCollationID > rCollationID) lCollationID else rCollationID
+    val collatorInfo = CollatorFactory.getInfoForId(collationID)
+    val collator = collatorInfo.collator
+    collator match {
+      // UTF8String collation
+      case null =>
+        collationID match {
+          // UCS_BASIC
+          case 0 =>
+            l.contains(r)
+          // UCS_BASIC_LCASE
+          case 1 =>
+            // scalastyle:off caselocale
+            l.toLowerCase.contains(r.toLowerCase)
+          // scalastyle:on caselocale
+          // Other special collations without ICU collator
+          case _ =>
+            var found = false
+            for (i <- 0 to l.numBytes() - r.numBytes() if !found) {
+              if (collatorInfo.comparator.compare(l.substring(i, i + r.numBytes()), r) == 0) {
+                found = true
+              }
+            }
+            found
```

Comment:
Note: for all collations that don't have a "collator", they have a "comparator", so the manual implementation works in a similar way as the one below (String ICU collation). So this code would work for UCS_BASIC (collationID=0) and UCS_BASIC_LCASE (collationID=1), but is probably less efficient than the above two implementations that use library functions.

Pro: this implementation is very similar to the one below, so it's "two birds with one stone" if we want to support all collation types at once.
Con: manual implementation takes more effort and requires more testing than using the library ones (where available).

Reply:
Update: These collations don't seem to exist at this point anyways, so I think this no longer makes sense. Removed in the new [PR](https://github.com/apache/spark/pull/45216/files)